### ISSUE TITLE
C# Null-coalescing operator is not correctly translated for constant DateTime value (DateTime.MinValue, DateTime.MaxValue)

### ISSUE THREAD

**Issue by laurent-jeancler-realist** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#1419]
### Steps to reproduce
Use Null-coalescing on a C# nullable date with fallback to constant DateTime.MinValue, DateTime.MaxValue.
Account.EndDate property is a nullable date time : DateTime?

Execute LINQ query :  `Accounts.Select(acc => acc.EndDate ?? DateTime.MinValue)`

### The issue
An InvalidCastException is thrown with message "Unable to cast object of type 'System.String' to type 'System.DateTime'."
```
System.InvalidCastException: Unable to cast object of type 'System.String' to type 'System.DateTime'.
   at MySqlConnector.Core.Row.GetDateTime(Int32 ordinal) in /_/src/MySqlConnector/Core/Row.cs:line 364
   at MySql.Data.MySqlClient.MySqlDataReader.GetDateTime(Int32 ordinal) in /_/src/MySqlConnector/MySql.Data.MySqlClient/MySqlDataReader.cs:line 251
   at lambda_method(Closure , QueryContext , DbDataReader , ResultContext , Int32[] , ResultCoordinator )
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.MoveNext()
   at LINQPad.ObjectGraph.ListNode.GetItems(IEnumerable list, Type staticDumpType, GraphOptions options)+MoveNext()
   at LINQPad.ObjectGraph.ListNode..ctor(ObjectNode parent, IEnumerable list, Type staticDumpType, GraphOptions options, String name)
   at LINQPad.ObjectGraph.ObjectNode.CreateInternal(ObjectNode parent, Object item, Type staticDumpType, GraphOptions options, Object itemForDescription, Boolean& collapse)
   at LINQPad.ObjectGraph.ObjectNode.TryCreate(ObjectNode parent, Object item, Type staticTypeDump, GraphOptions options)
   at LINQPad.ObjectGraph.ObjectNode.Create(Object item, Type staticDumpType, GraphOptions options)
   at LINQPad.ObjectGraph.Formatters.HtmlWriter.FormatObject(Object value, String heading, GraphOptions options, Boolean& inline, TimeSpan& formattingTime)
   at LINQPad.ObjectGraph.Formatters.OutputWriter.WriteObject(Object value, String heading, GraphOptions options, Boolean noNewLine)
   at LINQPad.ObjectGraph.Formatters.OutputWriter.WriteLineObject(Object value, String heading, GraphOptions options, Boolean noNewLine)
   at LINQPad.Extensions.DumpCore[T](T o, String description, Nullable`1 depth, Nullable`1 collapseTo, Boolean toDataGrid, String exclude, String include, Nullable`1 alpha, Boolean noTotals, Nullable`1 repeatHeadersAt, Boolean includePrivate)
   at LINQPad.Extensions.Dump[T](T o, String description, Nullable`1 depth, Nullable`1 collapseTo, Boolean toDataGrid, String exclude, String include, Nullable`1 alpha, Boolean noTotals, Nullable`1 repeatHeadersAt, Boolean includePrivate)
   at LINQPad.Extensions.Dump[T](T o)
   at UserQuery.Main() in C:\Users\XXXXXXXX\AppData\Local\Temp\LINQPad6\_rzfxrhmb\ytbqdm\LINQPadQuery:line 7
```

### The reason
This is because the translation of LINQ selector `acc.EndDate ?? DateTime.MinValue` generates faulty SQL : 
```
SELECT COALESCE(`a`.`end_date`, '0001-01-01 00:00:00')
FROM `accounts` AS `a`
```
The problem is the SQL COALESCE operator return type that is now a STRING !!!
So the DATASET contains a STRING field and EF is trying to read a DATETIME field to materialize the result, leading to the invalid cast exception.
### The solution
The correct translation must be something like this :
```
SELECT COALESCE(`a`.`end_date`, DATE('0001-01-01 00:00:00'))
FROM `accounts` AS `a`
```
### The workaround
Declaring a DateTime type variable to store the DateTime.MinValue, and using that variable in LINQ expression sovles the problem :
```
DateTime minValue = DateTime.MinValue;
var query = Accounts.Select(acc => acc.EndDate ?? minValue);
```
The way the query is now written uses the "minValue" parameter instead of translating DateTime.MinValue. The SQL COALESCE operator return type is now a DATE so EF materialzation works.

SQL :
```
-- Region Parameters
-- @__minValue_0='0001-01-01T00:00:00' (DbType = DateTime), @__p_1='10'
-- EndRegion
SELECT COALESCE(`a`.`end_date`, @__minValue_0)
FROM `accounts` AS `a`
```
### Further technical details

MySQL version: 8.0.21
Operating system: Windows 10
Pomelo.EntityFrameworkCore.MySql versions: 2.2.6, 3.2.4

**Comment by lauxjpn** [comment]
@laurent-jeancler-realist Thanks for the report.

We just fixed this a couple of days ago for our 6.0.0 Preview 4 release (e.g. see [MySqlDateTimeTypeMapping.cs](https://github.com/lauxjpn/Pomelo.EntityFrameworkCore.MySql/blob/2253976a4cfac99671180ddc4e2f007c8fef0b48/src/EFCore.MySql/Storage/Internal/MySqlDateTimeTypeMapping.cs#L75-L76)). I guess we can port it back to 5.0.0 as well.

As a quick workaround, it might also work to explicitly cast the result to a `DateTime`:

```c#
Accounts.Select(acc => (DateTime) (acc.EndDate ?? DateTime.MinValue))
```

**Comment by laurent-jeancler-realist** [comment]
Thank you, the fix is not needed for previous versions as the workaround I've posted works.

The explicit cast you mentioned was the first workaround I've tried, but it failed to translate as DATE(...) in SQL, the translated SQL remains :
```
SELECT COALESCE(`a`.`end_date`, '0001-01-01 00:00:00')
FROM `accounts` AS `a`
```
Regards,
